all: GraphQL API Versioning

kamilkisiela opened the pull request:
This is the initial implementation of versioning.

- [x] remove the example implementation of a feature flag

I left few comments on the interesting parts of the code to make it easier to review.
- [Where do we manage versions](https://github.com/graphprotocol/graph-node/pull/3185#discussion_r791502259)
- [How do we define "latest" version](https://github.com/graphprotocol/graph-node/pull/3185#discussion_r791502627)
- [How does the Gateway fetch a list of supported versions?](https://github.com/graphprotocol/graph-node/pull/3185#discussion_r791504128)
- [How do we define feature flags?](https://github.com/graphprotocol/graph-node/pull/3185#discussion_r791501883)
- [How do we validate the version number provided by the user?](https://github.com/graphprotocol/graph-node/pull/3185#discussion_r791503067)

Continues #3155

kamilkisiela commented on server/websocket/src/server.rs:
```diff
@@ -72,14 +76,18 @@ where
-                Ok(state(store, target_from_name(path_segments[1..].join("/"))).await)
+                Ok(state(store, target_from_id(subgraph_id, Default::default())).await)
             }
+            &["subgraphs", "name", _] | &["subgraphs", "name", _, _] => Ok(state(
+                store,
+                target_from_name(path_segments[2..].join("/"), Default::default()), // TODO: version
+            )
```
- [ ] resolve version of WebSocket connections

kamilkisiela commented on store/postgres/src/deployment_store.rs:
```diff
@@ -276,7 +277,9 @@ impl DeploymentStore {
         // if that's Fred the Dog, Fred the Cat or both.
         //
         // This assumes that there are no concurrent writes to a subgraph.
-        let schema = self.subgraph_info_with_conn(&conn, &layout.site)?.api;
+        let schema = self
+            .subgraph_info_with_conn(&conn, &layout.site, &Default::default())? // TODO: ask if we need a non-default version here
+            .api;
```
- [ ] Do we need a version here? Is default (latest) enough?

kamilkisiela commented on store/postgres/src/subgraph_store.rs:
```diff
@@ -676,11 +676,11 @@ impl SubgraphStoreInner {
         };
 
-        let (store, site) = self.store(&id)?;
+        let (store, site) = self.store(&id)?; // TODO: maybe use a tuple (id, version) here?
```
- [ ] not sure if we want to get the store based on `id` or `(id, version)`.

lutter replied:
Since the version only affects the API schema, I don't think there's a point in having the version here. I think version should only come into play for the `QueryStore`

kamilkisiela commented on store/postgres/src/subgraph_store.rs:
```diff
@@ -855,7 +855,7 @@ impl SubgraphStoreInner {
             let latest_ethereum_block_number =
                 chain.latest_block.as_ref().map(|ref block| block.number());
-            let subgraph_info = store.subgraph_info(site.as_ref())?;
+            let subgraph_info = store.subgraph_info(site.as_ref(), &Default::default())?; // TODO: ask if we need to use non-default version here
```
- [ ] Do we need a version here? Is default (latest) enough?

kamilkisiela commented on store/postgres/src/subgraph_store.rs:
```diff
@@ -1054,13 +1054,18 @@ impl SubgraphStoreTrait for SubgraphStore {
     fn input_schema(&self, id: &DeploymentHash) -> Result<Arc<Schema>, StoreError> {
         let (store, site) = self.store(&id)?;
-        let info = store.subgraph_info(site.as_ref())?;
+        // TODO: ask if we need to use non-default version here, I think we don't because it's schema provided by user
+        let info = store.subgraph_info(site.as_ref(), &Default::default())?;
```
- [ ] Do we need a version here? Is default (latest) enough?

I think we don't because it's schema provided by user

kamilkisiela commented on graph/src/components/versions/features.rs:
```diff
@@ -0,0 +1,5 @@
+#[derive(Clone, PartialEq)]
+pub enum FeatureFlag {
+    // A description of the feature. Give it a little context.
+    BasicOrdering,
```
An example flag, just to demo the versioning.

kamilkisiela commented on graph/src/components/versions/registry.rs:
```diff
@@ -0,0 +1,112 @@
+
+lazy_static! {
+    pub static ref VERSIONS: HashMap<VersionNumber, Vec<FeatureFlag>> = {
+        let supported_versions: Vec<(&str, Vec<FeatureFlag>)> = vec![
+            // baseline version
+            ("1.0.0", vec![]),
+            // Versions with feature flags
+            ("1.1.0", vec![FeatureFlag::BasicOrdering])
+        ];
```
This is something we need to maintain. It's the single source of truth.

kamilkisiela commented on graph/src/components/versions/registry.rs:
```diff
@@ -0,0 +1,112 @@
+        map
+    };
+
+    static ref LATEST_VERSION: String = {
+        let keys: Vec<VersionNumber> = VERSIONS.clone().into_keys().collect();
+
+        let last_version = keys.last().unwrap();
+
+        last_version.0.clone()
+    };
```
Latest version is the last key of `VERSIONS`.

kamilkisiela commented on graph/src/components/versions/registry.rs:
```diff
@@ -0,0 +1,112 @@
+}
+
+impl VersionNumber {
+    pub fn validate(version: String) -> Result<(), String> {
+        let chunks: Vec<&str> = version.split(".").collect();
+
+        if chunks.len() != 3 {
+            return Err(format!("Invalid version number: {}", version));
+        }
+
+        let major = chunks[0].parse::<u32>();
+        let minor = chunks[1].parse::<u32>();
+        let patch = chunks[2].parse::<u32>();
+
+        if major.is_err() || minor.is_err() || patch.is_err() {
+            return Err(format!("Invalid version number: {}", version));
+        }
+
+        if !VERSIONS.contains_key(&VersionNumber::from(version)) {
+            return Err("No versions found".to_string());
+        }
+
+        Ok(())
+    }
```
Validation happens in http and websocket part of the server.

kamilkisiela commented on graph/src/data/query/query.rs:
```diff
@@ -112,19 +112,20 @@ impl serde::ser::Serialize for QueryVariables {
 #[derive(Clone, Debug)]
 pub enum QueryTarget {
-    Name(SubgraphName),
-    Deployment(DeploymentHash),
+    Name(SubgraphName, VersionNumber),
+    Deployment(DeploymentHash, VersionNumber),
 }
```
I tried to look for the best place to put the requested version and `QueryTarget` won the contest.

kamilkisiela commented on graphql/src/schema/api.rs:
```diff
@@ -84,14 +84,20 @@ pub fn api_schema(input_schema: &Document) -> Result<Document, APISchemaError> {
     add_builtin_scalar_types(&mut schema)?;
-    add_order_direction_enum(&mut schema);
+
+    if version.supports(FeatureFlag::BasicOrdering) {
+        add_order_direction_enum(&mut schema);
+    }
```
An example implementation of a feature flag.

kamilkisiela commented on server/http/src/service.rs:
```diff
@@ -166,17 +166,45 @@ where
         self.serve_dynamic_file(self.graphiql_html())
     }
 
+    fn resolve_version_number(
+        &self,
+        request: &Request<Body>,
+    ) -> Result<VersionNumber, GraphQLServerError> {
+        let version_header = request.headers().get("api-version");
+
+        let mut version = VersionNumber::default();
+
+        if let Some(header_value) = version_header {
+            let version_value = header_value.to_str().map_err(|_| {
+                GraphQLServerError::ClientError(format!(
+                    "Invalid api-version header value {:?}",
+                    header_value
+                ))
+            })?;
+            VersionNumber::validate(version_value.to_string())
+                .map_err(|error| GraphQLServerError::ClientError(error))?;
+
+            version = version_value.into();
+        }
+
+        Ok(version)
+    }
```
I used `api-version` header because it was the easiest to implement quickly but before merging the Pull Request we should switch to url-based approach (this way we could use GraphiQL - it does not support headers).

kamilkisiela commented on server/index-node/src/schema.graphql:
```diff
@@ -19,6 +19,7 @@ type Query {
     indexer: Bytes
   ): Bytes
   subgraphFeatures(subgraphId: String!): SubgraphFeatures!
+  subgraphVersions(subgraphId: String!): [SubgraphVersion!]!
```
This is for the Gateway.

lutter replied:
Nice! We should avoid the words 'subgraph version' here since it already has a meaning in the context of `graph-node` as the mapping of subgraph names to deployments (IPFS hash) How about calling this `apiVersions`?  (That naming change should of course happen throughout the code supporting this GraphQL API)

kamilkisiela replied:
Done

kamilkisiela commented on server/index-node/src/schema.graphql:
```diff
@@ -95,3 +96,10 @@ enum Feature {
   fullTextSearch,
   ipfsOnEthereumContracts,
 }
+
+type SubgraphVersion {
+  """
+  Version number in SemVer format
+  """
+  version: String!
+}
```
Why an object instead of a list? It's extendable and harder to break backwards-compatibility.

kamilkisiela commented on server/websocket/src/connection.rs:
```diff
@@ -313,7 +313,7 @@ where
 
                     // Construct a subscription
-                    let target = QueryTarget::Deployment(deployment.clone());
+                    let target = QueryTarget::Deployment(deployment.clone(), Default::default()); // TODO: resolve and validate a version number from WebSocket connection
```
- [ ] resolve version of WebSocket connections

lutter replied:
I think we can leave this as-is since there are no subscriptions on the network anyway

kamilkisiela commented on graphql/src/schema/api.rs:
```diff
@@ -74,19 +74,20 @@ impl TryFrom<&r::Value> for ErrorPolicy {
-    add_field_arguments(&mut schema, input_schema)?;
-    add_query_type(&mut schema, &object_types, &interface_types)?;
-    add_subscription_type(&mut schema, &object_types, &interface_types)?;
+
```
```suggestion
```

evaporei commented:
@kamilkisiela is this ready to review? I see that the CI is failing.

kamilkisiela commented:
@evaporei yes, the code won't change much even after a rebase. I've got some questions both in code and gh comments, because I'm not sure if it's the right implementation.

lutter commented on store/postgres/src/deployment_store.rs:
```diff
@@ -87,7 +87,7 @@ pub struct StoreInner {
 
     /// A cache of commonly needed data about a subgraph.
-    subgraph_cache: Mutex<LruCache<DeploymentHash, SubgraphInfo>>,
+    subgraph_cache: Mutex<LruCache<(DeploymentHash, VersionNumber), SubgraphInfo>>,
```
How about keeping `DeploymentHash` as the cache key and changing `SubgraphInfo.api` to be a `HashMap<Version, Arc<ApiSchema>>`? As long as there's only a few API versions, we could even populate that `HashMap` eagerly; once there are more versions, we might want to populate entries there on demand.

That would make it possible to not pass in a version to `subgraph_info` and only use it when we actually need an `ApiSchema` in the `QueryStore`

kamilkisiela replied:
eaa5c19a8384bd804bc8b25aa11489e3ba2dabc1

kamilkisiela commented:
@azf20 no updates, I'm trying to implement support for interfaces in #3184 and once it's ready, I'll apply David's comments here.

kamilkisiela commented:
@lutter ready for a review again, I implemented everything you asked for. 

Those two failing tests (just the order of fields) are related to the fact I moved the ordering logic behind a feature flag to show what the versioning will look like. Once I remove the example feature flag, it will be fine.

kamilkisiela commented:
@azf20 I read the GIP and applied some changes to the PR. When we get a version range, the semver package resolves to the most recent applicable version.

lutter commented on graph/src/components/versions/registry.rs:
```diff
@@ -0,0 +1,87 @@
+        let mut map = HashMap::new();
+
+        // Sort version by major, minor, patch, from higher to lower.
+        supported_versions.sort_by(|a, b| b.0.partial_cmp(&a.0).unwrap());
```
Is this sort necessary? They all go into a `HashMap` afterwards, so doesn't seem like it's needed.

kamilkisiela replied:
I did a refactor, now only the vector is sorted

lutter commented on graph/src/components/versions/registry.rs:
```diff
@@ -0,0 +1,87 @@
+        let keys: Vec<Version> = VERSIONS.clone().into_keys().collect();
+
+        // Versions are sorted by major, minor, patch, from higher to lower.
+        keys.first().unwrap().to_string()
```
This is wrong `HashMap.into_keys` will return keys in arbitrary order. `BTreeMap` would work, but I think `first` would return the smallest version, so either reverse the iterator or put the keys into a `Vec` and sort that

kamilkisiela replied:
I did a refactor and it's now using `VERSION_COLLECTION: HashMap<Version, Vec<FeatureFlag>>` to get the feature flags and `VERSIONS: Vec<&'static Version>` (sorted from latest to oldest version, higher -> lower).

The `VERSIONS` vector is sorted because of the logic in `resolve()` function.

lutter commented on graph/src/components/versions/registry.rs:
```diff
@@ -0,0 +1,87 @@
+
+impl ApiVersion {
+    pub fn new(version_requirement: &VersionReq) -> Result<Self, String> {
+        let version = Self::resolve(&version_requirement);
```
It would be cleaner if `resolve` returned `Result<Self, String>` instead of an `Option`

kamilkisiela replied:
It does now, I did a refactor

lutter commented on graphql/src/schema/api.rs:
```diff
@@ -643,32 +655,36 @@ fn collection_arguments_for_named_type(type_name: &str) -> Vec<InputValue> {
+        ),
+    ];
+
+    if version.supports(FeatureFlag::BasicOrdering) {
```
I am losing track of the plot a little here - should we really have a version without ordering?

kamilkisiela replied:
No no, it's just an example of how to use the feature flags
https://github.com/graphprotocol/graph-node/pull/3185#discussion_r791501883

lutter reviewed (changes requested):
The main thing we need to clear up is what the initial version(s) should be

kamilkisiela commented:
@lutter I did a refactor, it includes your suggestions.

lutter reviewed (approved):
Looks good now!

@azf20 Are you ok with the base version being `1.0.0`? All fine by me, just wanted to check that that's where we should start.

azf20 commented:
Thanks @kamilkisiela and @lutter! I think starting with 1.0.0 is ok, but also interested in @That3Percent's expectations of a 1.x release

That3Percent commented:
> I think starting with 1.0.0 is ok, but also interested in @That3Percent's expectations of a 1.x release

I also think a 1.0.0 release is a good choice. Even if there are changes we know we would like to make, that's _always_ the case so it's better to have a 1.0 and 2.0 than 0.x in perpetuity.

That3Percent reviewed (commented):
Currently the protocol expects version "0" to exist so that we can be backward-compatible with the existing attestation format. As implemented, graph-node is only supporting 1.0.0 here.

I think this is probably fine, as long as we default to 1.0 upstream from here (indexer agent, gateway) and accept that version "0" should never be queried directly (it would be a historical relic). Thoughts?

azf20 commented:
Agree that this approach seems fine, and I think the GIP already allows for that kind of minimum version specification

azf20 commented:
Hey @kamilkisiela I think this is good to merge, though looks like there are a couple of conflicts first! cc @lutter
